backend/hydroshare_jupyter_sync/resource_hydroshare_data.py

```diff
@@ -0,0 +1,334 @@
+            :type dest_path: PosixPath
+        """
+        metadata = self.find_file_or_folder_metadata(str(src_path), self.get_files()["contents"])
+        if metadata["type"] == "folder":
+            for child_file_or_folder in metadata.get("contents"):
+                new_src = self.remove_prefix(child_file_or_folder.get("path"), HS_PREFIX + ':/')
+                self.rename_or_move_file(new_src, dest_path)
+        else:
+            self.hs.resource(self.res_id).functions.move_or_rename({
+                "source_path": str(src_path),
+                "target_path": str(dest_path),
+            })
+
+    def download_to_local(self, local_data, src_path, dest_path, temp_dir=None):
+        """ Copies a file or folder from HydroShare to the local filesystem
+
+            :param local_data: the local resource data
+            :type local_data: LocalFolder
+            :param src_path: the path to the file to copy/download (relative to the resource root)
+            :type src_path: PosixPath
+            :param dest_path: the path to copy the file to in the local resource (relative to the resource root)
+            :type dest_path: PosixPath
+            :param temp_dir: the temporary directory to use when downloading the file
+            :type temp_dir: PosixPath | None
+         """
+        metadata = self.find_file_or_folder_metadata(str(src_path), self.get_files()["contents"])
+        if metadata["type"] == "folder":
+            local_data.create_local_folder(dest_path)  # Does nothing if the folder already exists
+            for child_file_or_folder in metadata['contents']:
+                new_src = Path(self.remove_prefix(child_file_or_folder['path'], HS_PREFIX + ':/'))
+                self.download_to_local(local_data, new_src, dest_path / new_src.name, temp_dir)
+        else:
+            # Download the file
+            with open(str(local_data.data_path / dest_path), 'wb') as f:
+                for chunk in self.hs.getResourceFile(self.res_id, str(src_path)):
+                    f.write(chunk)
+
+    def upload_from_local(self, local_data, src_path, dest_path):
+        """ Copies a file or folder and its contents from the local filesystem to HydroShare
+
+            :param local_data: the local copy of the resource
+            :type local_data: LocalFolder
+            :param src_path: the path to the file or folder to upload (relative to the resource root)
+            :type src_path: PosixPath
+            :param dest_path: the path to the folder to upload to in the resource on HydroShare
+            :type dest_path: PosixPath
+        """
+        filesystem_item = local_data.data_path / src_path
+        if filesystem_item.is_dir():
+            try:  # Try to create the folder (it may already exist)
+                self.create_folder(dest_path)
+            except hs_restclient.HydroShareHTTPException:  # Folder already exists
```

Comment:
Just an FYI, when a folder already exists it throws and IndexError (not HydroshareHTTPException), which is currently being caught in the create_folder function.

Reply:
Didn't they fix that? Lemme check...

Reply:
Yeah, I'm pretty sure Scott fixed that in `hs_restclient`. This code works as intended.

Reply:
Wait I see your point now.

Reply:
OK yeah Scott did fix that. It never fails with an IndexError anymore (so I removed that part of the code below).

Reply:
lol did he really? I got an IndexError for an existing folder like 3 days ago so that was a very quick turnaround..

Reply:
So the IndexError was (afaik) a result of them calling the `HydroShareHTTPException` constructor and passing in the wrong sized tuple (and then accessing an element at a position that didn't exist). I believe they fixed those calls, which would have resolved this error. I only get `HydroShareHTTPException`s on my computer. I'd make sure your `hs_restclient` has been updated if that's not what you're getting.

Reply:
gotcha, yeah I guess I must be working with an outdated version of hs_restclient